File: library/json-mapper2/src/main/scala/com/anjunar/scala/schema/model/ObjectDescriptor.scala
package com.anjunar.scala.schema.model

import com.anjunar.scala.mapper.annotations.{PropertyDescriptor, IgnoreFilter}

import java.util
import scala.beans.BeanProperty
import scala.compiletime.uninitialized

@IgnoreFilter
class ObjectDescriptor extends NodeDescriptor {

  @PropertyDescriptor(title = "Properties")
  val properties : util.Map[String, NodeDescriptor] = new util.LinkedHashMap[String, NodeDescriptor]()

  @PropertyDescriptor(title = "One of")
  val oneOf : util.List[ObjectDescriptor] = new util.ArrayList[ObjectDescriptor]()
  
}

File: application/src/main/typescript/packages/react-simplicity/src/domain/descriptors/ObjectDescriptor.ts
import NodeDescriptor from "./NodeDescriptor";
import Basic from "../../mapper/annotations/Basic";
import Entity from "../../mapper/annotations/Entity";

interface PropertiesContainer {
    [key: string]: NodeDescriptor
}

@Entity("ObjectDescriptor")
export default class ObjectDescriptor extends NodeDescriptor {

    $type = "ObjectDescriptor"

    @Basic()
    properties: PropertiesContainer

    @Basic()
    oneOf : ObjectDescriptor[] = []

    allProperties(type : string) : PropertiesContainer {
        let objectDescriptor = this.oneOf.find(one => one.type === type);
        return Object.assign({}, this.properties, objectDescriptor?.properties)
    }

}